sdk/lib/rtl/amd64/unwind.c

```diff
@@ -687,3 +687,72 @@ RtlRaiseException(IN PEXCEPTION_RECORD ExceptionRecord)
     RtlRaiseStatus(Status);
 }
 
+static
+VOID
+RtlpCaptureNonVolatileContextPointers(
+    _Out_ PKNONVOLATILE_CONTEXT_POINTERS NonvolatileContextPointers,
+    _In_ ULONG64 TargetFrame)
+{
+    CONTEXT Context;
+    PRUNTIME_FUNCTION FunctionEntry;
+    ULONG64 ImageBase;
+    PVOID HandlerData;
+    ULONG64 EstablisherFrame;
+    KPROCESSOR_MODE Mode;
+
+    /* Zero out the nonvolatile context pointers */
+    RtlZeroMemory(NonvolatileContextPointers, sizeof(*NonvolatileContextPointers));
+
+    /* Capture the current context */
+    RtlCaptureContext(&Context);
+
+    /* Use the highest bit in RIP as mode flag */
+    Mode = Context.Rip >> 63;
+
+    do
+    {
+        /* Look up the function entry */
+        FunctionEntry = RtlLookupFunctionEntry(Context.Rip, &ImageBase, NULL);
+        ASSERT(FunctionEntry != NULL);
+
+        /* Do a virtual unwind to the caller */
+        RtlVirtualUnwind(UNW_FLAG_EHANDLER,
+                         ImageBase,
+                         Context.Rip,
+                         FunctionEntry,
+                         &Context,
+                         &HandlerData,
+                         &EstablisherFrame,
+                         NonvolatileContextPointers);
+
+        /* Continue until we reached the target frame or user mode */
+    } while ((EstablisherFrame < TargetFrame) && 
```

Comment:
Are we sure that the frame will always go upwards ? What if there is some strange stuff done with setjmp/longjmp ?

Reply:
longjmp is an unwind and actually goes through RtlUnwindEx (on Windows). Strange stuff with the stack is not allowed on x64.

Reply:
OK.